pkg/systemd/services/services.jsx

```diff
@@ -748,47 +596,82 @@ class ServicesPageBody extends React.Component {
         }
     }
 
-    render() {
-        const { unit_by_path } = this.state;
-        const path = this.props.path;
+    // compute filtered and sorted list of [Id, unit]
+    computeSelectedUnits() {
+        const unitType = '.' + this.props.activeTab;
+        const currentTextFilter = this.state.currentTextFilter.toLowerCase();
+        const filters = this.state.filters;
+        const selectedUnits = [];
+        const ids = new Set();
+
+        [...Object.entries(this.units), ...Object.entries(this.unit_files)].forEach(([idx, unit]) => {
+            if (!unit.Id?.endsWith(unitType))
+                return;
 
+            if (unit.LoadState === "not-found")
+                return;
+
+            // avoid showing unloaded units when there is a loaded one
+            if (ids.has(unit.Id))
+                return;
+            ids.add(unit.Id);
+
+            const UnitFileState = unit.UnitFileState ?? this.unit_files[unit.Id]?.UnitFileState;
+
+            if (currentTextFilter && !((unit.Description && unit.Description.toLowerCase().includes(currentTextFilter)) ||
+                unit.Id.toLowerCase().includes(currentTextFilter)))
+                return;
+
+            if (filters.fileState.length && this.unitFileState[UnitFileState] &&
+                !filters.fileState.includes(this.unitFileState[UnitFileState]))
+                return;
+
+            if (filters.activeState.length && this.activeState[unit.ActiveState] &&
+                !filters.activeState.includes(this.activeState[unit.ActiveState]))
+                return;
+
+            const augmentedUnit = { ...unit, UnitFileState, ...this.timers[idx] };
+            this.updateComputedProperties(augmentedUnit);
+            selectedUnits.push([unit.Id, augmentedUnit]);
+        });
+
+        selectedUnits.sort(this.compareUnits);
+        return selectedUnits;
+    }
+
+    componentDidUpdate(prevProps, prevState) {
```

Comment:
This seems to be triggered with a noticeable delay, half a second or so.  I think it is better to not rely on this at all and call computeSelectedUnits in render(), with a cache to prevent unnecessary recomputation.

Reply:
Hmm, no, that doesn't change anything visually.  There is still a noticeable delay between clicking on the service header and seeing the list update.  I am confused, more tomorrow.

Reply:
Interesting -- In Fedora 37 Chrome this is instantaneous for me. In firefox there's maybe 0.2s delay after I release the mouse button, it feels almost instant, and about the same delay that I get on main. Which browser are you using? Is that a remote machine? (Although that shouldn't matter -- switching tabs shouldn't do any D-Bus/bridge queries).

I'm a bit hesitant to drop this -- it's rather effective, and with the amount of redundant PropertyUpdates etc. that systemd throws at us we need some event coalescing. render() → computeSelectedUnits() → triggering a re-render sounds like asking for trouble?

Reply:
The delay does indeed come from constructing/processing a `ServicesList` with 310 entries.  Something outside of our own code seems to be busy for 500ms with that. When I change ServicesList to `return <>{units.map(u => <div>{u[0]}</div>)}</>`, then this delay disappears.

However, when I take this fast version of ServicesList and add a artificial delay of 2 seconds, it takes 4 seconds.

However, I think using `componentDidUpdate` and the "locationchanged" event is responsible for doubling the delay.  Here is the sequence of events when switching from "Targets" to "Services":

 - onLocationChanged() runs with `cockpit.locations.options == {type: "service"}`
 - it calls computeSelectedUnits() which still sees `this.props.activeTab == "target" and calls setState with that result.
 - usePageLocation triggers in the parent component and causes a re-render
 - render() of the Services component runs with `this.props.activeTab == "service"` but it renders `this.state.selectedUnits` which was computed when `activeTab` was still "target" (see above).
 - React goes into virtual DOM lala land for 500ms
 - componentDidUpdate runs and calls computeSelectedUnits, which sees `activeTab == "service" this time, and calls setState.
 - render() runs with the updated `this.state.selectedUnits`
 - React goes away into virtual DOM lala land again for 500ms.
 - The services list appears on screen, 1 second after the click.

A straight forward approach could be like this:

 - There is no state related to selected units or the filter parameters
 - render() calls computeSelectedUnits() each time and passes the return value straight to `ServicesList`.
 - computeSelectedUnits() looks at `this.props.activeTab` and `this.props.options` to find out the filter parameters (or render passes them in as parameters to computeSelectedUnits).

This will call computeSelectedUnits on each render. This is mostly fine, this is not the slow part by a mile.  If we want to optimize things, computeSelectedUnits can keep a memo cache of its most recent computation on `this`.

Reply:
Of course, a list with 300 entries should not take 500ms to render.  We might be doing something stupid there as well.